// git-filter-repo/lib/gitFilterRepo.js
require("core-js/modules/es.object.keys.js");

require("core-js/modules/es.symbol.js");

require("core-js/modules/es.array.filter.js");

require("core-js/modules/es.object.get-own-property-descriptor.js");

require("core-js/modules/es.object.get-own-property-descriptors.js");

(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "core-js/modules/es.array.index-of.js", "core-js/modules/web.dom-collections.for-each.js", "core-js/modules/es.array.concat.js", "core-js/modules/es.array.iterator.js", "core-js/modules/es.object.to-string.js", "core-js/modules/es.set.js", "core-js/modules/es.string.iterator.js", "core-js/modules/web.dom-collections.iterator.js", "core-js/modules/es.array.map.js", "core-js/modules/es.regexp.exec.js", "core-js/modules/es.string.split.js", "core-js/modules/es.string.replace.js", "@babel/runtime/regenerator", "@babel/runtime/helpers/newArrowCheck", "@babel/runtime/helpers/asyncToGenerator", "@babel/runtime/helpers/defineProperty", "@babel/runtime/helpers/classCallCheck", "@babel/runtime/helpers/createClass", "hash-wasm", "./git", "./gitDate", "./pip", "./socket"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("core-js/modules/es.array.index-of.js"), require("core-js/modules/web.dom-collections.for-each.js"), require("core-js/modules/es.array.concat.js"), require("core-js/modules/es.array.iterator.js"), require("core-js/modules/es.object.to-string.js"), require("core-js/modules/es.set.js"), require("core-js/modules/es.string.iterator.js"), require("core-js/modules/web.dom-collections.iterator.js"), require("core-js/modules/es.array.map.js"), require("core-js/modules/es.regexp.exec.js"), require("core-js/modules/es.string.split.js"), require("core-js/modules/es.string.replace.js"), require("@babel/runtime/regenerator"), require("@babel/runtime/helpers/newArrowCheck"), require("@babel/runtime/helpers/asyncToGenerator"), require("@babel/runtime/helpers/defineProperty"), require("@babel/runtime/helpers/classCallCheck"), require("@babel/runtime/helpers/createClass"), require("hash-wasm"), require("./git"), require("./gitDate"), require("./pip"), require("./socket"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.esArrayIndexOf, global.webDomCollectionsForEach, global.esArrayConcat, global.esArrayIterator, global.esObjectToString, global.esSet, global.esStringIterator, global.webDomCollectionsIterator, global.esArrayMap, global.esRegexpExec, global.esStringSplit, global.esStringReplace, global.regenerator, global.newArrowCheck, global.asyncToGenerator, global.defineProperty, global.classCallCheck, global.createClass, global.hashWasm, global.git, global.gitDate, global.pip, global.socket);
    global.gitFilterRepo = mod.exports;
  }
})(typeof globalThis !== "undefined" ? globalThis : typeof self !== "undefined" ? self : this, function (_exports, _esArrayIndexOf, _webDomCollectionsForEach, _esArrayConcat, _esArrayIterator, _esObjectToString, _esSet, _esStringIterator, _webDomCollectionsIterator, _esArrayMap, _esRegexpExec, _esStringSplit, _esStringReplace, _regenerator, _newArrowCheck2, _asyncToGenerator2, _defineProperty2, _classCallCheck2, _createClass2, _hashWasm, _git, _gitDate, _pip, _socket) {
  "use strict";

  var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  _exports.pythonCommitToCommit = pythonCommitToCommit;
  _exports.commitToPythonCommit = commitToPythonCommit;
  _exports.PathType = _exports.Operator = _exports.default = void 0;
  _regenerator = _interopRequireDefault(_regenerator);
  _newArrowCheck2 = _interopRequireDefault(_newArrowCheck2);
  _asyncToGenerator2 = _interopRequireDefault(_asyncToGenerator2);
  _defineProperty2 = _interopRequireDefault(_defineProperty2);
  _classCallCheck2 = _interopRequireDefault(_classCallCheck2);
  _createClass2 = _interopRequireDefault(_createClass2);
  _git = _interopRequireDefault(_git);
  _gitDate = _interopRequireDefault(_gitDate);
  _pip = _interopRequireDefault(_pip);
  _socket = _interopRequireDefault(_socket);

  function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

  function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

  var gitDate = new _gitDate.default();

  var GitFilterRepo = /*#__PURE__*/function () {
    function GitFilterRepo() {
      var gitPath = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : process.cwd();
      var options = arguments.length > 1 ? arguments[1] : undefined;
      (0, _classCallCheck2.default)(this, GitFilterRepo);
      this.gitPath = gitPath;
      this.git = void 0;
      this.pip = void 0;
      this.options = void 0;
      this.options = _objectSpread({
        pipe: true,
        preserveOrigin: false
      }, options);
      this.git = new _git.default(gitPath);
      this.pip = new _pip.default(gitPath);
    }

    (0, _createClass2.default)(GitFilterRepo, [{
      key: "ensure",
      value: function () {
        var _ensure = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
          var previousPath;
          return _regenerator.default.wrap(function _callee$(_context) {
            while (1) {
              switch (_context.prev = _context.next) {
                case 0:
                  previousPath = this.gitPath;
                  process.chdir(this.gitPath);
                  _context.next = 4;
                  return this.installed();

                case 4:
                  if (_context.sent) {
                    _context.next = 7;
                    break;
                  }

                  _context.next = 7;
                  return this.pip.install('git-filter-repo', {
                    user: true,
                    pipe: this.options.pipe
                  });

                case 7:
                  process.chdir(previousPath);

                case 8:
                case "end":
                  return _context.stop();
              }
            }
          }, _callee, this);
        }));

        function ensure() {
          return _ensure.apply(this, arguments);
        }

        return ensure;
      }()
    }, {
      key: "installed",
      value: function () {
        var _installed = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee2() {
          var _execaErr$stderr, execaErr;

          return _regenerator.default.wrap(function _callee2$(_context2) {
            while (1) {
              switch (_context2.prev = _context2.next) {
                case 0:
                  _context2.prev = 0;
                  _context2.next = 3;
                  return this.git.filterRepo({
                    help: true,
                    pipe: this.options.pipe
                  });

                case 3:
                  return _context2.abrupt("return", true);

                case 6:
                  _context2.prev = 6;
                  _context2.t0 = _context2["catch"](0);
                  execaErr = _context2.t0;

                  if (!(((_execaErr$stderr = execaErr.stderr) === null || _execaErr$stderr === void 0 ? void 0 : _execaErr$stderr.indexOf('is not a git command')) > -1)) {
                    _context2.next = 11;
                    break;
                  }

                  return _context2.abrupt("return", false);

                case 11:
                  throw _context2.t0;

                case 12:
                case "end":
                  return _context2.stop();
              }
            }
          }, _callee2, this, [[0, 6]]);
        }));

        function installed() {
          return _installed.apply(this, arguments);
        }

        return installed;
      }()
    }, {
      key: "invertPaths",
      value: function () {
        var _invertPaths = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee3(paths) {
          var _this = this;

          var options,
              filteredPaths,
              pathsGlob,
              pathsRegex,
              _args3 = arguments;
          return _regenerator.default.wrap(function _callee3$(_context3) {
            while (1) {
              switch (_context3.prev = _context3.next) {
                case 0:
                  options = _args3.length > 1 && _args3[1] !== undefined ? _args3[1] : {};
                  filteredPaths = [];
                  pathsGlob = [];
                  pathsRegex = [];
                  paths.forEach(function (path) {
                    (0, _newArrowCheck2.default)(this, _this);
                    if (!path) return;
                    var _ref = path,
                        type = _ref.type,
                        value = _ref.value;

                    if (type === PathType.Glob) {
                      pathsGlob.push(value);
                    } else if (type === PathType.Regex) {
                      pathsRegex.push(value);
                    } else if (typeof path === 'string' || value) {
                      paths.push(typeof path === 'string' ? path : value);
                    }
                  }.bind(this));
                  return _context3.abrupt("return", this.git.filterRepo(_objectSpread(_objectSpread({
                    force: true
                  }, options), {}, {
                    invertPaths: true,
                    paths: filteredPaths,
                    pathsGlob: pathsGlob,
                    pathsRegex: pathsRegex,
                    pipe: this.options.pipe
                  })));

                case 6:
                case "end":
                  return _context3.stop();
              }
            }
          }, _callee3, this);
        }));

        function invertPaths(_x) {
          return _invertPaths.apply(this, arguments);
        }

        return invertPaths;
      }()
    }, {
      key: "blobCallback",
      value: function () {
        var _blobCallback = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee5(callback) {
          var options,
              _args5 = arguments;
          return _regenerator.default.wrap(function _callee5$(_context5) {
            while (1) {
              switch (_context5.prev = _context5.next) {
                case 0:
                  options = _args5.length > 1 && _args5[1] !== undefined ? _args5[1] : {};
                  return _context5.abrupt("return", this.callback('blob', options, /*#__PURE__*/function () {
                    var _ref2 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee4(blob) {
                      return _regenerator.default.wrap(function _callee4$(_context4) {
                        while (1) {
                          switch (_context4.prev = _context4.next) {
                            case 0:
                              return _context4.abrupt("return", callback(blob));

                            case 1:
                            case "end":
                              return _context4.stop();
                          }
                        }
                      }, _callee4);
                    }));

                    return function (_x3) {
                      return _ref2.apply(this, arguments);
                    };
                  }()));

                case 2:
                case "end":
                  return _context5.stop();
              }
            }
          }, _callee5, this);
        }));

        function blobCallback(_x2) {
          return _blobCallback.apply(this, arguments);
        }

        return blobCallback;
      }()
    }, {
      key: "commitCallback",
      value: function () {
        var _commitCallback = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee7(callback) {
          var options,
              _args7 = arguments;
          return _regenerator.default.wrap(function _callee7$(_context7) {
            while (1) {
              switch (_context7.prev = _context7.next) {
                case 0:
                  options = _args7.length > 1 && _args7[1] !== undefined ? _args7[1] : {};
                  return _context7.abrupt("return", this.callback('commit', options, /*#__PURE__*/function () {
                    var _ref3 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee6(pythonCommit) {
                      return _regenerator.default.wrap(function _callee6$(_context6) {
                        while (1) {
                          switch (_context6.prev = _context6.next) {
                            case 0:
                              _context6.t0 = commitToPythonCommit;
                              _context6.next = 3;
                              return callback(pythonCommitToCommit(pythonCommit));

                            case 3:
                              _context6.t1 = _context6.sent;
                              return _context6.abrupt("return", (0, _context6.t0)(_context6.t1));

                            case 5:
                            case "end":
                              return _context6.stop();
                          }
                        }
                      }, _callee6);
                    }));

                    return function (_x5) {
                      return _ref3.apply(this, arguments);
                    };
                  }()));

                case 2:
                case "end":
                  return _context7.stop();
              }
            }
          }, _callee7, this);
        }));

        function commitCallback(_x4) {
          return _commitCallback.apply(this, arguments);
        }

        return commitCallback;
      }()
    }, {
      key: "tagCallback",
      value: function () {
        var _tagCallback = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee8(callback) {
          var _this2 = this;

          var options,
              _args8 = arguments;
          return _regenerator.default.wrap(function _callee8$(_context8) {
            while (1) {
              switch (_context8.prev = _context8.next) {
                case 0:
                  options = _args8.length > 1 && _args8[1] !== undefined ? _args8[1] : {};
                  return _context8.abrupt("return", this.callback('tag', options, function (tag) {
                    (0, _newArrowCheck2.default)(this, _this2);
                    return callback(tag);
                  }.bind(this)));

                case 2:
                case "end":
                  return _context8.stop();
              }
            }
          }, _callee8, this);
        }));

        function tagCallback(_x6) {
          return _tagCallback.apply(this, arguments);
        }

        return tagCallback;
      }()
    }, {
      key: "resetCallback",
      value: function () {
        var _resetCallback = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee10(callback) {
          var options,
              _args10 = arguments;
          return _regenerator.default.wrap(function _callee10$(_context10) {
            while (1) {
              switch (_context10.prev = _context10.next) {
                case 0:
                  options = _args10.length > 1 && _args10[1] !== undefined ? _args10[1] : {};
                  return _context10.abrupt("return", this.callback('reset', options, /*#__PURE__*/function () {
                    var _ref4 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee9(reset) {
                      return _regenerator.default.wrap(function _callee9$(_context9) {
                        while (1) {
                          switch (_context9.prev = _context9.next) {
                            case 0:
                              return _context9.abrupt("return", callback(reset));

                            case 1:
                            case "end":
                              return _context9.stop();
                          }
                        }
                      }, _callee9);
                    }));

                    return function (_x8) {
                      return _ref4.apply(this, arguments);
                    };
                  }()));

                case 2:
                case "end":
                  return _context10.stop();
              }
            }
          }, _callee10, this);
        }));

        function resetCallback(_x7) {
          return _resetCallback.apply(this, arguments);
        }

        return resetCallback;
      }()
    }, {
      key: "filenameCallback",
      value: function () {
        var _filenameCallback = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee12(callback) {
          var options,
              _args12 = arguments;
          return _regenerator.default.wrap(function _callee12$(_context12) {
            while (1) {
              switch (_context12.prev = _context12.next) {
                case 0:
                  options = _args12.length > 1 && _args12[1] !== undefined ? _args12[1] : {};
                  return _context12.abrupt("return", this.callback('filename', options, /*#__PURE__*/function () {
                    var _ref5 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee11(filename) {
                      return _regenerator.default.wrap(function _callee11$(_context11) {
                        while (1) {
                          switch (_context11.prev = _context11.next) {
                            case 0:
                              return _context11.abrupt("return", callback(filename));

                            case 1:
                            case "end":
                              return _context11.stop();
                          }
                        }
                      }, _callee11);
                    }));

                    return function (_x10) {
                      return _ref5.apply(this, arguments);
                    };
                  }()));

                case 2:
                case "end":
                  return _context12.stop();
              }
            }
          }, _callee12, this);
        }));

        function filenameCallback(_x9) {
          return _filenameCallback.apply(this, arguments);
        }

        return filenameCallback;
      }()
    }, {
      key: "messageCallback",
      value: function () {
        var _messageCallback = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee14(callback) {
          var options,
              _args14 = arguments;
          return _regenerator.default.wrap(function _callee14$(_context14) {
            while (1) {
              switch (_context14.prev = _context14.next) {
                case 0:
                  options = _args14.length > 1 && _args14[1] !== undefined ? _args14[1] : {};
                  return _context14.abrupt("return", this.callback('message', options, /*#__PURE__*/function () {
                    var _ref6 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee13(message) {
                      return _regenerator.default.wrap(function _callee13$(_context13) {
                        while (1) {
                          switch (_context13.prev = _context13.next) {
                            case 0:
                              return _context13.abrupt("return", callback(message));

                            case 1:
                            case "end":
                              return _context13.stop();
                          }
                        }
                      }, _callee13);
                    }));

                    return function (_x12) {
                      return _ref6.apply(this, arguments);
                    };
                  }()));

                case 2:
                case "end":
                  return _context14.stop();
              }
            }
          }, _callee14, this);
        }));

        function messageCallback(_x11) {
          return _messageCallback.apply(this, arguments);
        }

        return messageCallback;
      }()
    }, {
      key: "nameCallback",
      value: function () {
        var _nameCallback = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee16(callback) {
          var options,
              _args16 = arguments;
          return _regenerator.default.wrap(function _callee16$(_context16) {
            while (1) {
              switch (_context16.prev = _context16.next) {
                case 0:
                  options = _args16.length > 1 && _args16[1] !== undefined ? _args16[1] : {};
                  return _context16.abrupt("return", this.callback('name', options, /*#__PURE__*/function () {
                    var _ref7 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee15(name) {
                      return _regenerator.default.wrap(function _callee15$(_context15) {
                        while (1) {
                          switch (_context15.prev = _context15.next) {
                            case 0:
                              return _context15.abrupt("return", callback(name));

                            case 1:
                            case "end":
                              return _context15.stop();
                          }
                        }
                      }, _callee15);
                    }));

                    return function (_x14) {
                      return _ref7.apply(this, arguments);
                    };
                  }()));

                case 2:
                case "end":
                  return _context16.stop();
              }
            }
          }, _callee16, this);
        }));

        function nameCallback(_x13) {
          return _nameCallback.apply(this, arguments);
        }

        return nameCallback;
      }()
    }, {
      key: "emailCallback",
      value: function () {
        var _emailCallback = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee18(callback) {
          var options,
              _args18 = arguments;
          return _regenerator.default.wrap(function _callee18$(_context18) {
            while (1) {
              switch (_context18.prev = _context18.next) {
                case 0:
                  options = _args18.length > 1 && _args18[1] !== undefined ? _args18[1] : {};
                  return _context18.abrupt("return", this.callback('email', options, /*#__PURE__*/function () {
                    var _ref8 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee17(email) {
                      return _regenerator.default.wrap(function _callee17$(_context17) {
                        while (1) {
                          switch (_context17.prev = _context17.next) {
                            case 0:
                              return _context17.abrupt("return", callback(email));

                            case 1:
                            case "end":
                              return _context17.stop();
                          }
                        }
                      }, _callee17);
                    }));

                    return function (_x16) {
                      return _ref8.apply(this, arguments);
                    };
                  }()));

                case 2:
                case "end":
                  return _context18.stop();
              }
            }
          }, _callee18, this);
        }));

        function emailCallback(_x15) {
          return _emailCallback.apply(this, arguments);
        }

        return emailCallback;
      }()
    }, {
      key: "refnameCallback",
      value: function () {
        var _refnameCallback = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee20(callback) {
          var options,
              _args20 = arguments;
          return _regenerator.default.wrap(function _callee20$(_context20) {
            while (1) {
              switch (_context20.prev = _context20.next) {
                case 0:
                  options = _args20.length > 1 && _args20[1] !== undefined ? _args20[1] : {};
                  return _context20.abrupt("return", this.callback('refname', options, /*#__PURE__*/function () {
                    var _ref9 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee19(refname) {
                      return _regenerator.default.wrap(function _callee19$(_context19) {
                        while (1) {
                          switch (_context19.prev = _context19.next) {
                            case 0:
                              return _context19.abrupt("return", callback(refname));

                            case 1:
                            case "end":
                              return _context19.stop();
                          }
                        }
                      }, _callee19);
                    }));

                    return function (_x18) {
                      return _ref9.apply(this, arguments);
                    };
                  }()));

                case 2:
                case "end":
                  return _context20.stop();
              }
            }
          }, _callee20, this);
        }));

        function refnameCallback(_x17) {
          return _refnameCallback.apply(this, arguments);
        }

        return refnameCallback;
      }()
    }, {
      key: "callback",
      value: function () {
        var _callback2 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee21(name) {
          var _this3 = this,
              _objectSpread2;

          var options,
              _callback,
              socket,
              hasOriginRemote,
              result,
              _args21 = arguments;

          return _regenerator.default.wrap(function _callee21$(_context21) {
            while (1) {
              switch (_context21.prev = _context21.next) {
                case 0:
                  options = _args21.length > 1 && _args21[1] !== undefined ? _args21[1] : {};
                  _callback = _args21.length > 2 && _args21[2] !== undefined ? _args21[2] : function () {
                    (0, _newArrowCheck2.default)(this, _this3);
                    return null;
                  }.bind(this);
                  _context21.next = 4;
                  return this.ensure();

                case 4:
                  socket = new _socket.default('captain_hook', (0, _defineProperty2.default)({}, "".concat(name, "Callback"), _callback));
                  _context21.next = 7;
                  return socket.connect();

                case 7:
                  _context21.t0 = this.options.preserveOrigin;

                  if (!_context21.t0) {
                    _context21.next = 12;
                    break;
                  }

                  _context21.next = 11;
                  return this.hasRemote('origin');

                case 11:
                  _context21.t0 = _context21.sent;

                case 12:
                  hasOriginRemote = _context21.t0;

                  if (!hasOriginRemote) {
                    _context21.next = 30;
                    break;
                  }

                  _context21.t1 = this.git;
                  _context21.t2 = {
                    pipe: this.options.pipe
                  };
                  _context21.next = 18;
                  return (0, _hashWasm.xxhash3)('origin');

                case 18:
                  _context21.t3 = _context21.sent;
                  _context21.t4 = ['remove', _context21.t3];
                  _context21.next = 22;
                  return _context21.t1.remote.call(_context21.t1, _context21.t2, _context21.t4).catch(function () {
                    (0, _newArrowCheck2.default)(this, _this3);
                    return null;
                  }.bind(this));

                case 22:
                  _context21.t5 = this.git;
                  _context21.t6 = {
                    pipe: this.options.pipe
                  };
                  _context21.next = 26;
                  return (0, _hashWasm.xxhash3)('origin');

                case 26:
                  _context21.t7 = _context21.sent;
                  _context21.t8 = ['rename', 'origin', _context21.t7];
                  _context21.next = 30;
                  return _context21.t5.remote.call(_context21.t5, _context21.t6, _context21.t8);

                case 30:
                  _context21.next = 32;
                  return this.git.filterRepo(_objectSpread(_objectSpread({
                    force: true
                  }, options), {}, (_objectSpread2 = {}, (0, _defineProperty2.default)(_objectSpread2, "".concat(name, "Callback"), "return callbacks.callback('".concat(name, "', ").concat(name, ")")), (0, _defineProperty2.default)(_objectSpread2, "pipe", this.options.pipe), (0, _defineProperty2.default)(_objectSpread2, "importScripts", ['callbacks']), _objectSpread2)));

                case 32:
                  result = _context21.sent;

                  if (!hasOriginRemote) {
                    _context21.next = 42;
                    break;
                  }

                  _context21.t9 = this.git;
                  _context21.t10 = {
                    pipe: this.options.pipe
                  };
                  _context21.next = 38;
                  return (0, _hashWasm.xxhash3)('origin');

                case 38:
                  _context21.t11 = _context21.sent;
                  _context21.t12 = ['rename', _context21.t11, 'origin'];
                  _context21.next = 42;
                  return _context21.t9.remote.call(_context21.t9, _context21.t10, _context21.t12);

                case 42:
                  _context21.next = 44;
                  return socket.close();

                case 44:
                  return _context21.abrupt("return", result);

                case 45:
                case "end":
                  return _context21.stop();
              }
            }
          }, _callee21, this);
        }));

        function callback(_x19) {
          return _callback2.apply(this, arguments);
        }

        return callback;
      }()
    }, {
      key: "help",
      value: function () {
        var _help = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee22() {
          var options,
              _args22 = arguments;
          return _regenerator.default.wrap(function _callee22$(_context22) {
            while (1) {
              switch (_context22.prev = _context22.next) {
                case 0:
                  options = _args22.length > 0 && _args22[0] !== undefined ? _args22[0] : {};
                  return _context22.abrupt("return", this.git.filterRepo(_objectSpread(_objectSpread({}, options), {}, {
                    help: true,
                    pipe: this.options.pipe
                  })));

                case 2:
                case "end":
                  return _context22.stop();
              }
            }
          }, _callee22, this);
        }));

        function help() {
          return _help.apply(this, arguments);
        }

        return help;
      }()
    }, {
      key: "hasRemote",
      value: function () {
        var _hasRemote = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee23(remote) {
          var _this4 = this;

          return _regenerator.default.wrap(function _callee23$(_context23) {
            while (1) {
              switch (_context23.prev = _context23.next) {
                case 0:
                  _context23.t0 = Set;
                  _context23.next = 3;
                  return this.git.remote({
                    pipe: this.options.pipe
                  }, '-v');

                case 3:
                  _context23.t1 = _context23.sent.split('\n').map(function (remote) {
                    (0, _newArrowCheck2.default)(this, _this4);
                    return remote.replace(/\s+.*$/g, '');
                  }.bind(this));
                  return _context23.abrupt("return", new _context23.t0(_context23.t1).has(remote));

                case 5:
                case "end":
                  return _context23.stop();
              }
            }
          }, _callee23, this);
        }));

        function hasRemote(_x20) {
          return _hasRemote.apply(this, arguments);
        }

        return hasRemote;
      }()
    }]);
    return GitFilterRepo;
  }();

  _exports.default = GitFilterRepo;
  var Operator;
  _exports.Operator = Operator;

  (function (Operator) {
    Operator["Equal"] = "=";
    Operator["GreaterThan"] = ">";
    Operator["GreaterThanOrEqual"] = ">=";
    Operator["LessThan"] = "<";
    Operator["LessThanOrEqual"] = "<=";
    Operator["NotEqual"] = "!=";
  })(Operator || (_exports.Operator = Operator = {}));

  function pythonCommitToCommit(pythonCommit) {
    return {
      authorDate: gitDate.dateFromGitDate(pythonCommit.author_date),
      authorEmail: pythonCommit.author_email,
      authorName: pythonCommit.author_name,
      branch: pythonCommit.branch,
      committerDate: gitDate.dateFromGitDate(pythonCommit.committer_date),
      committerEmail: pythonCommit.committer_email,
      committerName: pythonCommit.committer_name,
      dumped: pythonCommit.dumped,
      id: pythonCommit.id,
      message: pythonCommit.message,
      oldId: pythonCommit.old_id,
      originalId: pythonCommit.original_id,
      type: pythonCommit.type
    };
  }

  function commitToPythonCommit(commit) {
    return {
      author_date: gitDate.gitDateFromDate(commit.authorDate),
      author_email: commit.authorEmail,
      author_name: commit.authorName,
      branch: commit.branch,
      committer_date: gitDate.gitDateFromDate(commit.committerDate),
      committer_email: commit.committerEmail,
      committer_name: commit.committerName,
      dumped: commit.dumped,
      id: commit.id,
      message: commit.message,
      old_id: commit.oldId,
      original_id: commit.originalId,
      type: commit.type
    };
  }

  var PathType;
  _exports.PathType = PathType;

  (function (PathType) {
    PathType["Glob"] = "glob";
    PathType["Regex"] = "regex";
  })(PathType || (_exports.PathType = PathType = {}));
});
