// git-filter-repo/lib/index.js
var _typeof = require("@babel/runtime/helpers/typeof");

require("core-js/modules/web.dom-collections.for-each.js");

require("core-js/modules/es.object.keys.js");

require("core-js/modules/es.array.iterator.js");

require("core-js/modules/es.object.to-string.js");

require("core-js/modules/es.string.iterator.js");

require("core-js/modules/es.weak-map.js");

require("core-js/modules/web.dom-collections.iterator.js");

require("core-js/modules/es.object.get-own-property-descriptor.js");

(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "./git", "./gitDate", "./gitFilterRepo", "./pip", "./socket", "./types"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("./git"), require("./gitDate"), require("./gitFilterRepo"), require("./pip"), require("./socket"), require("./types"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.git, global.gitDate, global.gitFilterRepo, global.pip, global.socket, global.types);
    global.index = mod.exports;
  }
})(typeof globalThis !== "undefined" ? globalThis : typeof self !== "undefined" ? self : this, function (_exports, _git, _gitDate, _gitFilterRepo, _pip, _socket, _types) {
  "use strict";

  var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

  Object.defineProperty(_exports, "__esModule", {
    value: true
  });
  var _exportNames = {
    gitDate: true,
    Git: true,
    GitDate: true,
    Pip: true,
    Socket: true
  };
  Object.defineProperty(_exports, "Git", {
    enumerable: true,
    get: function get() {
      return _git.default;
    }
  });
  Object.defineProperty(_exports, "GitDate", {
    enumerable: true,
    get: function get() {
      return _gitDate.default;
    }
  });
  Object.defineProperty(_exports, "Pip", {
    enumerable: true,
    get: function get() {
      return _pip.default;
    }
  });
  Object.defineProperty(_exports, "Socket", {
    enumerable: true,
    get: function get() {
      return _socket.default;
    }
  });
  _exports.default = _exports.gitDate = void 0;
  _git = _interopRequireWildcard(_git);
  Object.keys(_git).forEach(function (key) {
    if (key === "default" || key === "__esModule") return;
    if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
    if (key in _exports && _exports[key] === _git[key]) return;
    Object.defineProperty(_exports, key, {
      enumerable: true,
      get: function get() {
        return _git[key];
      }
    });
  });
  _gitDate = _interopRequireDefault(_gitDate);
  _gitFilterRepo = _interopRequireWildcard(_gitFilterRepo);
  Object.keys(_gitFilterRepo).forEach(function (key) {
    if (key === "default" || key === "__esModule") return;
    if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
    if (key in _exports && _exports[key] === _gitFilterRepo[key]) return;
    Object.defineProperty(_exports, key, {
      enumerable: true,
      get: function get() {
        return _gitFilterRepo[key];
      }
    });
  });
  _pip = _interopRequireWildcard(_pip);
  Object.keys(_pip).forEach(function (key) {
    if (key === "default" || key === "__esModule") return;
    if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
    if (key in _exports && _exports[key] === _pip[key]) return;
    Object.defineProperty(_exports, key, {
      enumerable: true,
      get: function get() {
        return _pip[key];
      }
    });
  });
  _socket = _interopRequireWildcard(_socket);
  Object.keys(_socket).forEach(function (key) {
    if (key === "default" || key === "__esModule") return;
    if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
    if (key in _exports && _exports[key] === _socket[key]) return;
    Object.defineProperty(_exports, key, {
      enumerable: true,
      get: function get() {
        return _socket[key];
      }
    });
  });
  Object.keys(_types).forEach(function (key) {
    if (key === "default" || key === "__esModule") return;
    if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
    if (key in _exports && _exports[key] === _types[key]) return;
    Object.defineProperty(_exports, key, {
      enumerable: true,
      get: function get() {
        return _types[key];
      }
    });
  });

  function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

  function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

  /**
   * Copyright 2021 Silicon Hills LLC
   *
   * Licensed under the Apache License, Version 2.0 (the "License");
   * you may not use this file except in compliance with the License.
   * You may obtain a copy of the License at
   *
   * http://www.apache.org/licenses/LICENSE-2.0
   *
   * Unless required by applicable law or agreed to in writing, software
   * distributed under the License is distributed on an "AS IS" BASIS,
   * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   * See the License for the specific language governing permissions and
   * limitations under the License.
   */
  var gitDate = new _gitDate.default();
  _exports.gitDate = gitDate;
  var _default = _gitFilterRepo.default;
  _exports.default = _default;
});
