// git-filter-repo/es/pip.js
import "core-js/modules/es.object.keys.js";
import "core-js/modules/es.symbol.js";
import "core-js/modules/es.array.filter.js";
import "core-js/modules/es.object.get-own-property-descriptor.js";
import "core-js/modules/web.dom-collections.for-each.js";
import "core-js/modules/es.object.get-own-property-descriptors.js";
import _toConsumableArray from "@babel/runtime/helpers/toConsumableArray";
import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _regeneratorRuntime from "@babel/runtime/regenerator";
import "core-js/modules/es.array.concat.js";
import "core-js/modules/es.array.join.js";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * Copyright 2021 Silicon Hills LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import execa from 'execa';

var Pip = /*#__PURE__*/function () {
  function Pip() {
    var gitPath = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : process.cwd();
    var logger = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : console;

    _classCallCheck(this, Pip);

    this.gitPath = gitPath;
    this.logger = logger;
  }

  _createClass(Pip, [{
    key: "install",
    value: function () {
      var _install = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee(packageName) {
        var options,
            _user$options,
            user,
            _args = arguments;

        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                options = _args.length > 1 && _args[1] !== undefined ? _args[1] : {};
                _user$options = _objectSpread({
                  user: true
                }, options), user = _user$options.user;
                delete options.user;
                return _context.abrupt("return", this.run(['install'].concat(_toConsumableArray(user ? ['--user'] : []), [packageName]), options));

              case 4:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function install(_x) {
        return _install.apply(this, arguments);
      }

      return install;
    }()
  }, {
    key: "run",
    value: function () {
      var _run = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2(args) {
        var _p$stdout;

        var options,
            _dryrun$pipe$cwd$opti,
            cwd,
            pipe,
            dryrun,
            argsArr,
            command,
            p,
            _yield$p,
            stdout,
            _args2 = arguments;

        return _regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                options = _args2.length > 1 && _args2[1] !== undefined ? _args2[1] : {};
                _dryrun$pipe$cwd$opti = _objectSpread({
                  dryrun: false,
                  pipe: false,
                  cwd: process.cwd()
                }, options), cwd = _dryrun$pipe$cwd$opti.cwd, pipe = _dryrun$pipe$cwd$opti.pipe, dryrun = _dryrun$pipe$cwd$opti.dryrun;
                argsArr = _toConsumableArray(Array.isArray(args) ? args : [args]);
                command = "pip ".concat(argsArr.join(' '));

                if (!dryrun) {
                  _context2.next = 7;
                  break;
                }

                this.logger.info(command);
                return _context2.abrupt("return", command);

              case 7:
                p = execa('sh', ['-c', command], {
                  stdio: 'pipe',
                  cwd: cwd
                });
                if (pipe) (_p$stdout = p.stdout) === null || _p$stdout === void 0 ? void 0 : _p$stdout.pipe(process.stdout);
                _context2.next = 11;
                return p;

              case 11:
                _yield$p = _context2.sent;
                stdout = _yield$p.stdout;
                _context2.prev = 13;
                return _context2.abrupt("return", JSON.parse(stdout));

              case 17:
                _context2.prev = 17;
                _context2.t0 = _context2["catch"](13);
                return _context2.abrupt("return", stdout);

              case 20:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this, [[13, 17]]);
      }));

      function run(_x2) {
        return _run.apply(this, arguments);
      }

      return run;
    }()
  }]);

  return Pip;
}();

export { Pip as default };
